Add correlation cost layer

PatWie opened the pull request:
I just copied the files from my PR in tensorflow and slightly (+ blindly) adjusted the `BUILD`. Hence, this is really WIP.

References:
- https://github.com/tensorflow/tensorflow/pull/21392#issuecomment-440720577

I will rebase to master when I can run those ci test locally + adding a comprehensive commit message.
The docker command seems to not contain any way to run CUDA.

PatWie commented:
I just copied over the file and slightly (+ blindly) adjusted the `BUILD`. Hence, this is really WIP.

seanpmorgan commented:
For reference: 
https://github.com/tensorflow/tensorflow/pull/21392

facaiy commented:
The GPU support is still in process: https://github.com/tensorflow/addons/issues/118

PatWie commented:
I did a rebase to the master branch and fixed some path+lines in my code to adapt it to this addons-repo.

I **will not** join the *tf-eager-execution-club*. I added the `tf.compat.v1.disable_eager_execution()` to have a proper TensorFlow without this eager-exec-bull***t to preserve my own sanity and let other people take care of the tf-eager-crap.

seanpmorgan commented:
Lol thanks you're entitled to your opinion. I would like to work on this, but will probably wait until dynamic kernels are finalized. Once it's done maybe we can benchmark autograph and graph mode performance of the same layer. 

I'll reach back out to you if/when I have questions if thats alright.

seanpmorgan commented:
Just glanced at the code it's all the op call it seems so probably will not need to benchmark. I'll see if we can adapt it to eager in the short term. GPU support might wait for dynamic kernels so we don't package 2 addons

PatWie commented:
The GPU support is already in the BUILD for BAZEL (in the comments). I can do the adjustments for the GPU stuff when dynamic kernels with Cuda is described somewhere.
The unit-tests are quite slow for the CPU version.

Honestly, I tried the eager part for the tests, but I had no time to find a good way to compare CPU and GPU output.

Btw: The workflow from the contribution.md in this repo with the docker image is nice and helpful!

olesalscheider commented on tensorflow_addons/custom_ops/opticalflow/BUILD:
```diff
@@ -0,0 +1,26 @@
+    copts = [
+        "-pthread",
+        "-std=c++11",
+        "-D_GLIBCXX_USE_CXX11_ABI=0",
```
Should this really be hardcoded here?

kyleabeauchamp replied:
We have a few open tickets about how we're going to refactor this config (e.g., https://github.com/tensorflow/addons/pull/211) to a more central location.  However, it is still unresolved.

seanpmorgan replied:
The ABI? No it shouldn't but we'll be discussing how we want bazel to set this. See #211

PatWie replied:
I don't like it. But it is a copy of the other BUILD files from this repo. My default would be "1" anyway

olesalscheider commented:
Any news on this?

seanpmorgan commented:
This got pushed for some other work, but GPU testing is now setup so I'll try to have this done by the end of the week

seanpmorgan commented:
@tensorflow/sig-addons-maintainers Ready for review.

Tests are run in eager and graph, exposed as a Keras Layer and op if needed. GPU compiling / testing works.

There is a fail on py2-gpu that seems unrelated to this. It's the same fail from todays CI run and looks like its a bug in tf-core nightly or something.

PatWie commented on tensorflow_addons/layers/optical_flow.py:
```diff
@@ -0,0 +1,207 @@
+        self.stride_1 = stride_1
+        self.stride_2 = stride_2
+        self.pad = pad
+        self.data_format = data_format
```
Is see no check for possible data-formats `channels_first`. Is this caught somewhere else?

seanpmorgan replied:
There was a check in the op, but I agree better to catch at layer construction.

PatWie commented on tensorflow_addons/layers/optical_flow_test.py:
```diff
@@ -0,0 +1,199 @@
+
+    def _forward_simple(self, data_format):
+        # cumbersome calculation by hand for a fixed input
+        # we just test where zeros occurs and a few entries
```
I know that I wrote it. But I would change it to
> We are just testing where the output has vanishing values.

PatWie commented on tensorflow_addons/layers/optical_flow_test.py:
```diff
@@ -0,0 +1,199 @@
+                    pad=pad,
+                    data_format=data_format)
+
+            theoretical, numerical = tf.test.compute_gradient(
```
I would rename `theoretical` to `gradient_from_implementation` and `numerical` to `numerical_gradient`.

seanpmorgan replied:
I do like that re-name, but going to leave it so its consistent with TF Core:
https://github.com/tensorflow/tensorflow/search?utf8=%E2%9C%93&q=theoretical%2C+numerical&type=

PatWie commented on tensorflow_addons/layers/optical_flow_test.py:
```diff
@@ -0,0 +1,199 @@
+            self.assertAllClose(theoretical[0], numerical[0], atol=1e-3)
+
+    def _keras(self, data_format):
+        # Unable to use `layer_test` as this layer has multiple inputs
```
missing `.`

facaiy commented:
@PatWie Hi, Patrick. Could you add some docstring for this class when time allows? I find its API document lacks necessary information: 
https://github.com/tensorflow/addons/blob/r0.5/docs/api_docs/python/tfa/layers/CorrelationCost.md

It seems that we should hide correlation_cost function, and only expose CorrelationCost in public.
https://github.com/tensorflow/addons/blob/r0.5/docs/api_docs/python/tfa/layers/optical_flow/correlation_cost.md

Please correct me if I'm wrong, thank you :-)

PatWie commented:
@facaiy . The code is documented (just the c++ part)! Does 

https://github.com/tensorflow/addons/blob/76249547c481ed47bd6f2f8092a99cd64d37fd2f/tensorflow_addons/custom_ops/layers/cc/ops/correlation_cost_op.cc#L92-L109

help? Like I already stated, I'm not a big friend of Keras, TF2 and co. I create my graphs as I did since the first TF has been release.

facaiy commented:
Unfortunately, I'm afraid that we only provide API document in python side for users. No worries, I filed an issue to seek help from community https://github.com/tensorflow/addons/issues/509

PyExtreme commented:
Hi @PatWie , Since the documentation has already been done in C++ part. I am just copying and pasting it.
As per another task in this issue, I will make the correlation_cost function private.

@seanpmorgan , Please let me know if any other changes are there. I am new to open source and will appreciate your inputs.

Thanks

WindQAQ commented:
Hi @PyExtreme, many thanks for the help! What you mentioned above are the points for this issue. Just go ahead so that we could directly give you some feedback during review! Thanks again :smiley:
